rfd/0094-discover-metrics.md

```diff
@@ -0,0 +1,195 @@
+  use to instrument other parts of Teleport and website.
+* The events should not include any customer identifiable information.
+
+## Details
+
+We want to be able to answer the following question to analyze the full Discover
+flow. Below, by "resource" we mean a server, Kubernetes cluster, database,
+application or Windows desktop.
+
+* _Once the user signed up, how long did it take them to add a first resource?_
+  It's the question we can answer today and, as discussed above, not very useful
+  by itself but still needed for a full picture.
+
+* _How long did the actual process of adding a first resource took?_ Meaning,
+  from the moment they started going through the Discover Wizard (a datapoint
+  we lack today) to the successful completion.
+
+* _How long does adding a resource take?_ It's a more generic version of the
+  above, which is a general indicator of how easy it is to add new resource.
+
+* _Which steps of adding a new resource take most time?_ Discover is a multi-
+  step wizard and this will help understand where users get stuck the most.
+
+* _Which types of resources/protocols take most time?_ To be able to pinpoint,
+  for example, which databases users have most issues with.
+
+* _Where do users experience failures when adding resources?_ For example,
+  agent does not join in time or connection tester fails.
+
+Most of the questions we're interested in can be answered using PostHog
+[funnels](https://posthog.com/manual/funnels) as they are designed to help
+understand how users progress through the product flows and where the
+bottlenecks and friction are.
+
+To be able to build a Discover funnel, each step the user goes through needs
+to be defined as an event. Below is a Protobuf spec containing proposed events
+metadata. Note that each event will include anonymized cluster name and timestamp
+as shown in our [reporting spec](https://github.com/gravitational/prehog/blob/main/proto/prehog/v1alpha/teleport.proto)
+so below we only include event metadata.
+
+```protobuf
+// Contains common metadata for Discover related events.
+message DiscoverMetadata {
+    // Uniquely identifies Discover wizard "session". Will allow to correlate
+    // events within the same Discover wizard run. Can be UUIDv4.
+    string id = 1;
+}
+
+// Contains common metadata identifying resource type being added.
+message ResourceMetadata {
+    // Resource type: server, kubernetes, database, app, desktop.
+    string type = 1;
+    // Optional protocol where applicable, e.g. postgres, mysql, sqlserver.
+    string protocol = 2;
+}
+
+// Contains fields that track a particular step outcome, for example connection
+// test failed or succeeded, or user aborted the step.
+message Status {
+    // Indicates the step outcome. For example, "success" means user proceeded
+    // to the next step, "error" if something failed (e.g. connection test),
+    // "aborted" if user exited the wizard.
+    string status = 1;
```

Comment:
I think we should use Enum here
Should be to normalize the query values in Posthog

Reply:
GRPC is just our internal communication format for these events, when converted to Posthog events these fields would become properties. But I updated to use enum anyway.